core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -540,11 +543,28 @@ private void updateCacheStream(long blockId) throws IOException {
     }
 
     try {
-      WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
-          mBlockStore.getWorkerInfoList(), getBlockSizeAllocation(mPos));
-      // If we reach here, we need to cache.
-      mCurrentCacheStream =
-          mBlockStore.getOutStream(blockId, getBlockSize(mPos), address, mOutStreamOptions);
+      List<BlockWorkerInfo> workers = mBlockStore.getWorkerInfoList();
+      // If we are reading a block from a remote worker, we shouldn't cache back to it.
+      if (mCurrentBlockInStream instanceof RemoteBlockInStream) {
+        InetSocketAddress address = ((RemoteBlockInStream) mCurrentBlockInStream).location();
+        BlockWorkerInfo remoteWorker = null;
+        for (BlockWorkerInfo worker : workers) {
+          if (worker.getNetAddress().getHost().equals(address.getHostString())
+              && worker.getNetAddress().getDataPort() == address.getPort()) {
+            remoteWorker = worker;
+          }
+        }
```

Comment:
Can you use `iterator` and then `remove`? This does not work if there are multiple entries for the same worker (though that should never happen).